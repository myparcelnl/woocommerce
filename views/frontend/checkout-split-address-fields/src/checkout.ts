/* eslint-disable @typescript-eslint/naming-convention */

jQuery(document).ready(($) => {
  /* Hide custom NL fields by default when country not NL */
  const billingCountry = $('#billing_country').val();
  const shippingCountry = $('#shipping_country').val();

  if (billingCountry != 'NL') {
    $('#billing_street_name_field').hide();
    $('#billing_house_number_field').hide();
    $('#billing_house_number_suffix_field').hide();
  }

  if (shippingCountry != 'NL') {
    $('#shipping_street_name_field').hide();
    $('#shipping_house_number_field').hide();
    $('#shipping_house_number_suffix_field').hide();
  }

  /* Localisation */
  const localeJson = woocommerce_params.locale.replace(/&quot;/g, '"');
  const locale = $.parseJSON(localeJson);
  const required = ` <abbr class="required" title="${woocommerce_params.i18n_required_text}">*</abbr>`;

  $('body')
    /* Handle locale */
    .bind('country_to_state_changing', (event, country, wrapper) => {
      let thislocale;
      const thisform = wrapper;
      const $postcodefield = thisform.find('#billing_postcode_field, #shipping_postcode_field');
      const $cityfield = thisform.find('#billing_city_field, #shipping_city_field');
      const $address1field = thisform.find('#billing_address_1_field, #shipping_address_1_field');
      const $address2field = thisform.find('#billing_address_2_field, #shipping_address_2_field');
      const $streetfield = thisform.find('#billing_street_name_field, #shipping_street_name_field');
      const $numberfield = thisform.find('#billing_house_number_field, #shipping_house_number_field');
      const $suffixfield = thisform.find('#billing_house_number_suffix_field, #shipping_house_number_suffix_field');

      if (country == 'NL') {
        /* show custom NL fields */
        $streetfield.show();
        $numberfield.show();
        $suffixfield.show();

        /* Hide regular address classes */
        $address1field.find('label abbr').remove();
        $address1field.hide();
        $address2field.hide();

        /* Place postcode & city on one line */
        $postcodefield.removeClass('form-row-wide').addClass('form-row-first');
        $cityfield.removeClass('form-row-wide').addClass('form-row-last');

        /* Mark required fields */
        if ($streetfield.find('label abbr').size() == 0) {
          $streetfield.find('label').append(required);
        }

        if ($numberfield.find('label abbr').size() == 0) {
          $numberfield.find('label').append(required);
        }

        /* Add validation required classes */
        $streetfield.addClass('validate-required');
        $numberfield.addClass('validate-required');
      } else {
        /* Hide custom NL fields */
        $streetfield.hide();
        $numberfield.hide();
        $suffixfield.hide();

        /* Unmark required fields */
        $streetfield.find('label abbr').remove();
        $numberfield.find('label abbr').remove();

        /* Remove validation required classes */
        $streetfield.removeClass('validate-required');
        $numberfield.removeClass('validate-required');
      }

      if (typeof locale[country] != 'undefined') {
        thislocale = locale[country];
      } else {
        thislocale = locale.default;
      }

      /* Handle locale fields */
      const locale_fields = {
        address_1: '#billing_address_1_field, #shipping_address_1_field',
        address_2: '#billing_address_2_field, #shipping_address_2_field',
      };

      $.each(locale_fields, function (key, value) {
        const field = thisform.find(value);

        if (thislocale[key]) {
          if (thislocale[key].label) {
            field.find('label').html(thislocale[key].label);
          }

          if (thislocale[key].placeholder) {
            field.find('input').attr('placeholder', thislocale[key].placeholder);
          }

          field.find('label abbr').remove();

          if (typeof thislocale[key].required == 'undefined' && locale.default[key].required == true) {
            field.find('label').append(required);
          } else if (thislocale[key].required == true) {
            field.find('label').append(required);
          }

          if (key !== 'state') {
            if (thislocale[key].hidden == true) {
              field.hide().find('input').val('');
            } else {
              field.show();
            }
          }
        } else if (locale.default[key]) {
          if (locale.default[key].required == true) {
            if (field.find('label abbr').size() == 0) {
              field.find('label').append(required);
            }
          }

          if (key !== 'state' &&
            (typeof locale.default[key].hidden == 'undefined' || locale.default[key].hidden == false)
          ) {
            field.show();
          }
        }
      });
    })

    /* Init trigger */
    .bind('init_checkout', function () {
      $('#billing_country, #shipping_country, .country_to_state').change();
    });

  /* Update on page load */
  if (woocommerce_params.is_checkout == 1) {
    $('body').trigger('init_checkout');
  }
});
