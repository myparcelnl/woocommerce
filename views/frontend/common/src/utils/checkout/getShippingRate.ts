import {type WcShippingRate} from '../../types';
import {useCartStore} from './useCartStore';

export const getShippingRate = (): WcShippingRate | undefined => {
  const cartStore = useCartStore();
  const shippingRates = cartStore.getShippingRates();

  return shippingRates[0].shipping_rates.find((rate) => rate.selected);
};
