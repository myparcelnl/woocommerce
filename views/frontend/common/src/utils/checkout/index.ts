export * from './getBlocksCheckoutConfig';
export * from './getClassicCheckoutConfig';
export * from './getShippingRate';
export * from './isClassicCheckout';
export * from './useWcCartStore';
export * from './useWcCheckoutStore';
