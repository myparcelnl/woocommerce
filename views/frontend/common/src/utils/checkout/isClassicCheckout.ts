export const isClassicCheckout = (): boolean => {
  return document.querySelector('form[name="checkout"]') !== null;
};
