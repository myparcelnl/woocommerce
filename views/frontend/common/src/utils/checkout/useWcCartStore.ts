import {type StoreDescriptor} from '@wordpress/data';
import {type WcCartStore, type UseWcStore} from '../../types';

export const useWcCartStore = (): UseWcStore<WcCartStore> => {
  const {CART_STORE_KEY} = window.wc.wcBlocksData as {CART_STORE_KEY: StoreDescriptor};
  return {
    storeDescriptor: CART_STORE_KEY,
    selectors: wp.data.select(CART_STORE_KEY),
    actions: wp.data.dispatch(CART_STORE_KEY),
  };
};
