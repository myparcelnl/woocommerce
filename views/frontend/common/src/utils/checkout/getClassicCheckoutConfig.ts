import {AddressType, useConfig} from '@myparcel-pdk/checkout-common';
import {useUtil, AddressField, PdkUtil, SeparateAddressField} from '@myparcel-pdk/checkout';
import {type CheckoutConfig} from '../../types';

// eslint-disable-next-line max-lines-per-function
export const getClassicCheckoutConfig = (): CheckoutConfig => {
  return {
    addressFields: {
      [AddressField.Address1]: `address_1`,
      [AddressField.Address2]: `address_2`,
      [AddressField.City]: `city`,
      [AddressField.Country]: `country`,
      [AddressField.PostalCode]: `postcode`,
      [SeparateAddressField.Street]: `street_name`,
      [SeparateAddressField.Number]: `house_number`,
      [SeparateAddressField.NumberSuffix]: `house_number_suffix`,
    },

    prefixBilling: 'billing_',
    prefixShipping: 'shipping_',

    fieldShippingMethod: 'shipping_method',
    shippingMethodFormField: 'shipping_method[0]',

    config: {
      formChange(callback) {
        jQuery(this.getForm()).on('change', () => {
          callback();
        });
      },

      getForm() {
        const getElement = useUtil(PdkUtil.GetElement);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return getElement('form[name="checkout"]')!;
      },

      getFormData() {
        const form = useConfig().getForm();
        const formData = new FormData(form);

        return Object.fromEntries(formData.entries());
      },

      hasAddressType(addressType: AddressType) {
        const billingElement = document.querySelector('.woocommerce-billing-fields__field-wrapper');

        return AddressType.Shipping === addressType || billingElement !== null;
      },

      initialize() {
        return new Promise((resolve) => {
          jQuery(() => {
            resolve();
          });
        });
      },
    },
  } satisfies CheckoutConfig;
};
