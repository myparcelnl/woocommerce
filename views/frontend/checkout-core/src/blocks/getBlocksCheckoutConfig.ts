import {FIELD_PREFIX_BILLING, FIELD_PREFIX_SHIPPING} from '@myparcel-woocommerce/frontend-common';
import {AddressType, PdkField, type PdkCheckoutConfigInput, useConfig} from '@myparcel-pdk/checkout-common';
import {PdkUtil, useUtil, type PdkFormData, updateContext, useCheckoutStore} from '@myparcel-pdk/checkout';
import {useCartStore, getShippingRate, createFields} from '../utils';
import {SELECTOR_FORM_BLOCKS} from '../constants';
import {updateAddressInStore} from './updateAddressInStore';
import {resolveAddress} from './resolveAddress';
import {createBlocksCheckoutFieldSelectors} from './createBlocksCheckoutFieldSelectors';
import {ADDRESS_FIELDS_BLOCKS_CHECKOUT} from './constants';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types,max-lines-per-function
export const getBlocksCheckoutConfig = () => {
  return {
    fields: {
      // Irrelevant because there's always a separate billing and shipping address object.
      [PdkField.AddressType]: '',
      // Value is manually resolved in getFormData.
      [PdkField.ShippingMethod]: '',
      [AddressType.Billing]: createBlocksCheckoutFieldSelectors(`${FIELD_PREFIX_BILLING}-`),
      [AddressType.Shipping]: createBlocksCheckoutFieldSelectors(`${FIELD_PREFIX_SHIPPING}-`),
    },

    formData: {
      [PdkField.AddressType]: PdkField.AddressType,
      [PdkField.ShippingMethod]: PdkField.ShippingMethod,
      [AddressType.Billing]: createFields(ADDRESS_FIELDS_BLOCKS_CHECKOUT, (val) => `${FIELD_PREFIX_BILLING}-${val}`),
      [AddressType.Shipping]: createFields(ADDRESS_FIELDS_BLOCKS_CHECKOUT, (val) => `${FIELD_PREFIX_SHIPPING}-${val}`),
    },

    /**
     * Update whenever the shipping method or the address changes.
     */
    formChange(callback) {
      const cartStore = useCartStore();

      let previousShippingRate = getShippingRate();
      let previousCustomerData = JSON.stringify(cartStore.getCustomerData());

      wp.data.subscribe(async () => {
        const currentShippingRate = getShippingRate();
        const currentCustomerData = cartStore.getCustomerData();

        const shippingMethodChanged = previousShippingRate?.rate_id !== currentShippingRate?.rate_id;
        const customerDataChanged = previousCustomerData !== JSON.stringify(currentCustomerData);

        if (!shippingMethodChanged && !customerDataChanged) {
          return;
        }

        if (customerDataChanged) {
          previousCustomerData = JSON.stringify(currentCustomerData);
        }

        if (shippingMethodChanged) {
          previousShippingRate = currentShippingRate;

          await updateContext();
        }

        callback();
      });
    },

    // eslint-disable-next-line max-lines-per-function
    getFormData() {
      const cartStore = useCartStore();
      const {billingAddress, shippingAddress} = cartStore.getCustomerData();
      const formData: PdkFormData = {};

      [AddressType.Shipping, AddressType.Billing].forEach((addressType) => {
        const addressObject = addressType === AddressType.Billing ? billingAddress : shippingAddress;
        const newAddress = resolveAddress(addressObject);

        Object.entries(newAddress).forEach(([key, value]) => {
          formData[`${addressType}-${key}`] = value;
        });
      });

      const shippingRates = cartStore.getShippingRates();
      const selectedRate = shippingRates[0].shipping_rates.find((rate) => rate.selected);

      return {
        ...formData,
        [PdkField.AddressType]: AddressType.Shipping,
        [PdkField.ShippingMethod]: selectedRate?.rate_id,
      };
    },

    getForm() {
      const getElement = useUtil(PdkUtil.GetElement);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      return getElement<HTMLFormElement>(SELECTOR_FORM_BLOCKS)!;
    },

    hasAddressType() {
      return true;
    },

    initialize() {
      return new Promise((resolve) => {
        document.addEventListener('myparcel_wc_delivery_options_ready', () => {
          const config = useConfig();
          const form = config.getForm();

          form.addEventListener('change', (...args) => {
            const checkoutStore = useCheckoutStore();

            updateAddressInStore(AddressType.Shipping, resolveAddress(checkoutStore.state.form[AddressType.Shipping]));
            updateAddressInStore(AddressType.Billing, resolveAddress(checkoutStore.state.form[AddressType.Billing]));
          });

          resolve();
        });
      });
    },
  } satisfies Partial<PdkCheckoutConfigInput>;
};
