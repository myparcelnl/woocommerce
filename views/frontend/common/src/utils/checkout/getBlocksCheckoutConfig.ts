import {AddressField, AddressType, PdkField, type PdkFormData} from '@myparcel-pdk/checkout-common';
import {PdkUtil, useUtil, updateContext, SeparateAddressField} from '@myparcel-pdk/checkout';
import {type CheckoutConfig} from '../../types';
import {useWcCartStore} from './useWcCartStore';
import {getShippingRate} from './getShippingRate';

const MYPARCEL_BLOCK_FIELDS_PREFIX = 'myparcelnl/';

// eslint-disable-next-line max-lines-per-function
export const getBlocksCheckoutConfig = (): CheckoutConfig => {
  const addressFields = {
    eoriNumber: `eori_number`,
    vatNumber: `vat_number`,
    [AddressField.Address1]: `address_1`,
    [AddressField.Address2]: `address_2`,
    [AddressField.City]: `city`,
    [AddressField.Country]: `country`,
    [AddressField.PostalCode]: `postcode`,
    [SeparateAddressField.Street]: `${MYPARCEL_BLOCK_FIELDS_PREFIX}street_name`,
    [SeparateAddressField.Number]: `${MYPARCEL_BLOCK_FIELDS_PREFIX}house_number`,
    [SeparateAddressField.NumberSuffix]: `${MYPARCEL_BLOCK_FIELDS_PREFIX}house_number_suffix`,
  };

  const hasAddressType = (addressType: AddressType): boolean => {
    const billingElement = document.querySelector('#billing-fields');

    return AddressType.Shipping === addressType || billingElement !== null;
  };

  return {
    addressFields,
    fieldAddressType: 'checkbox-control-0',
    fieldShippingMethod: 'shipping_method',
    prefixBilling: 'billing-',
    prefixShipping: 'shipping-',

    shippingMethodFormDataKey: PdkField.ShippingMethod,
    addressTypeFormDataKey: PdkField.AddressType,

    config: {
      /**
       * Update whenever the shipping method or the address changes.
       */
      formChange(callback) {
        const wcCartStore = useWcCartStore();
        let previousShippingRate = getShippingRate();
        let previousCustomerData = JSON.stringify(wcCartStore.selectors.getCustomerData());

        wp.data.subscribe(async () => {
          const currentShippingRate = getShippingRate();
          const currentCustomerData = wcCartStore.selectors.getCustomerData();

          const shippingMethodChanged = previousShippingRate?.rate_id !== currentShippingRate?.rate_id;
          const customerDataChanged = previousCustomerData !== JSON.stringify(currentCustomerData);

          if (!shippingMethodChanged && !customerDataChanged) {
            return;
          }

          if (customerDataChanged) {
            previousCustomerData = JSON.stringify(currentCustomerData);
          }

          if (shippingMethodChanged) {
            previousShippingRate = currentShippingRate;

            await updateContext();
          }

          callback();
        });
      },

      getFormData() {
        const wcCartStore = useWcCartStore();
        const customerData = wcCartStore.selectors.getCustomerData();
        const formData: PdkFormData = {};

        [AddressType.Shipping, AddressType.Billing].forEach((addressType) => {
          Object.keys(addressFields).forEach((field) => {
            formData[`${addressType}-${addressFields[field]}`] =
              customerData[`${addressType}Address`][addressFields[field]];
          });
        });

        const shippingRates = wcCartStore.selectors.getShippingRates();
        const selectedRate = shippingRates[0]?.shipping_rates.find((rate) => rate.selected);

        formData[PdkField.ShippingMethod] = selectedRate?.rate_id;

        // In the blocks checkout, the shipping address is *always* shown and billing address is optional. MyParcel uses the shipping address only.
        formData[PdkField.AddressType] = AddressType.Shipping;
        return formData;
      },

      getAddressType() {
        // Always use shipping in the blocks checkout.
        return AddressType.Shipping;
      },

      getForm() {
        const getElement = useUtil(PdkUtil.GetElement);

        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return getElement('.wc-block-checkout__form')!;
      },

      hasAddressType,

      initialize() {
        return new Promise((resolve) => {
          document.addEventListener('myparcel_wc_delivery_options_ready', () => {
            resolve();
          });
        });
      },
    },
  } satisfies CheckoutConfig;
};
