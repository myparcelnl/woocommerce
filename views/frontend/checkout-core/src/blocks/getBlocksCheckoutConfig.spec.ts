import {describe, it, expect} from 'vitest';
import {AddressType} from '@myparcel-pdk/checkout';
import {getBlocksCheckoutConfig} from './getBlocksCheckoutConfig';

/**
 * @vitest-environment happy-dom
 */

describe('getBlocksCheckoutConfig', () => {
  it('has correct fields and formData config', () => {
    const {fields, formData} = getBlocksCheckoutConfig();

    expect(fields[AddressType.Billing]).toMatchSnapshot();
    expect(fields[AddressType.Shipping]).toMatchSnapshot();

    expect(formData[AddressType.Billing]).toMatchSnapshot();
    expect(formData[AddressType.Shipping]).toMatchSnapshot();
  });
});
