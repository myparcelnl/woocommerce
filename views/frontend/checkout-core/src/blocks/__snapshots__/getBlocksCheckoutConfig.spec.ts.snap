// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getBlocksCheckoutConfig > has correct fields and formData config 1`] = `
{
  "address1": "#billing-address_1",
  "address2": "#billing-address_2",
  "city": "#billing-city",
  "country": "#billing-country",
  "eoriNumber": "#billing-myparcelnl-eori_number",
  "number": "#billing-myparcelnl-house_number",
  "numberSuffix": "#billing-myparcelnl-house_number_suffix",
  "postalCode": "#billing-postcode",
  "street": "#billing-myparcelnl-street_name",
  "vatNumber": "#billing-myparcelnl-vat_number",
}
`;

exports[`getBlocksCheckoutConfig > has correct fields and formData config 2`] = `
{
  "address1": "#shipping-address_1",
  "address2": "#shipping-address_2",
  "city": "#shipping-city",
  "country": "#shipping-country",
  "eoriNumber": "#shipping-myparcelnl-eori_number",
  "number": "#shipping-myparcelnl-house_number",
  "numberSuffix": "#shipping-myparcelnl-house_number_suffix",
  "postalCode": "#shipping-postcode",
  "street": "#shipping-myparcelnl-street_name",
  "vatNumber": "#shipping-myparcelnl-vat_number",
}
`;

exports[`getBlocksCheckoutConfig > has correct fields and formData config 3`] = `
{
  "address1": "billing-address_1",
  "address2": "billing-address_2",
  "city": "billing-city",
  "country": "billing-country",
  "eoriNumber": "billing-myparcelnl/eori_number",
  "number": "billing-myparcelnl/house_number",
  "numberSuffix": "billing-myparcelnl/house_number_suffix",
  "postalCode": "billing-postcode",
  "street": "billing-myparcelnl/street_name",
  "vatNumber": "billing-myparcelnl/vat_number",
}
`;

exports[`getBlocksCheckoutConfig > has correct fields and formData config 4`] = `
{
  "address1": "shipping-address_1",
  "address2": "shipping-address_2",
  "city": "shipping-city",
  "country": "shipping-country",
  "eoriNumber": "shipping-myparcelnl/eori_number",
  "number": "shipping-myparcelnl/house_number",
  "numberSuffix": "shipping-myparcelnl/house_number_suffix",
  "postalCode": "shipping-postcode",
  "street": "shipping-myparcelnl/street_name",
  "vatNumber": "shipping-myparcelnl/vat_number",
}
`;
